Add ThroughputMonitor for enhanced data transfer tracking

t-davidbrown-msft opened the pull request:
Introduces the ThroughputMonitor class to monitor and calculate throughput during data transfer operations.

Updates
- JobBuilder,
- TransferJobInternal,
- TransferManager, and
- TransferProgressTracker to integrate throughput monitoring.

Adds unit tests for ThroughputMonitor functionality and updates existing tests to mock and verify its integration in transfer processes.

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

Commit:
Add ThroughputMonitor for enhanced data transfer tracking

Introduces the ThroughputMonitor class to monitor and calculate throughput during data transfer operations.

Updates
- JobBuilder,
- TransferJobInternal,
- TransferManager, and
- TransferProgressTracker to integrate throughput monitoring.

Adds unit tests for ThroughputMonitor functionality and updates existing tests to mock and verify its integration in transfer processes.

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Storage.DataMovement](https://apiview.dev/Assemblies/Review/fffb907bd4224ac4b0f95dfb164a72e3?revisionId=744e4924a0c54176a3d4551e2414bf36)

Commit:
Refactor ThroughputMonitor constructor

added a new parameterless constructor that initializes the instance with a default processor.

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Storage.DataMovement](https://apiview.dev/Assemblies/Review/fffb907bd4224ac4b0f95dfb164a72e3?revisionId=744e4924a0c54176a3d4551e2414bf36)

Commit:
change interval for throughput monitor tests

Commit:
change public class to internal class

Commit:
Fix ThroughputMonitor instantiation in TransferManager constructor

jalauzon-msft commented on sdk/storage/Azure.Storage.DataMovement/src/ConcurrencyTuner/ThroughputMonitor.cs:
```diff
@@ -0,0 +1,153 @@
+            if (_totalBytesTransferred == 0)
+                return Task.CompletedTask;
+
+            if (!_isStopwatchRunning)
```
Is there a way we can get rid of stopwatch?

t-davidbrown-msft replied:
@jalauzon-msft done.

jalauzon-msft commented on sdk/storage/Azure.Storage.DataMovement/src/ConcurrencyTuner/ThroughputMonitor.cs:
```diff
@@ -0,0 +1,153 @@
+    /// Monitors the throughput of data transfer operations by tracking the total bytes transferred
+    /// and calculating the throughput in bytes per second.
+    /// </summary>
+    internal class ThroughputMonitor : IAsyncDisposable
```
This can be an `IDiposable`

t-davidbrown-msft replied:
@jalauzon-msft done.

amnguye commented on sdk/storage/Azure.Storage.DataMovement/src/ConcurrencyTuner/ThroughputMonitor.cs:
```diff
@@ -0,0 +1,153 @@
+        /// </summary>
+        /// <param name="bytesTransferred">The number of bytes transferred.</param>
+        /// <param name="_">A cancellation token (not used).</param>
+        private Task ProcessBytesTransferredAsync(long bytesTransferred, CancellationToken _)
```
Probably should check if cancellationtoken was cancelled at the start of the method.

`CancellationHelper.ThrowIfCancellationRequested(token);`

t-davidbrown-msft replied:
I think @jalauzon-msft mentioned this a while back. He didn't want the CancellationToken to cancel the entire throughput monitor.

Commit:
Refactor TransferManager for parameterized options

Significantly modify the TransferManager class to replace the
previous TransferOptions property with a parameterized approach
using transferOptions in various methods. Update ResumeAllTransfersAsync
and ResumeTransferAsync to utilize the transferOptions parameter,
ensuring robustness by initializing transferOptions when null.
Enhance TransferProgressTracker to guarantee proper initialization
of ThroughputMonitor and improve progress tracking consistency.

Commit:
Remove comment in TransferManager

Commit:
Remove TransferOptions from TransferManager class

The `TransferOptions` property and its associated private field `_transferOptions` have been removed from the `TransferManager` class. This change eliminates the property implementation that managed the assignment and validation of transfer options, including the handling of `ProgressHandlerOptions`. This simplification may reflect a shift in how transfer options are managed or a refactoring of the codebase.

Commit:
Improve throughput calculations and test coverage

- Modify `AvgThroughput` to return `double.MaxValue` for zero elapsed time with bytes transferred.
- Update `ProcessBytesTransferredAsync` to handle zero elapsed time correctly.
- Change `DisposeAsync` to a synchronous `Dispose` method.
- Refactor and rename tests in `ThroughputMonitorTests.cs` for clarity.
- Add new tests for `AvgThroughput` under various conditions.
- Enhance tests to verify internal state using reflection.
- Clarify cancellation token handling in `QueueBytesTransferredAsync`.

Commit:
Update throughput test assertions with tolerance value

Modified the `ThroughputInMb_And_AvgThroughputInMb_CalculatedCorrectly` test in `ThroughputMonitorTests.cs` to use a tolerance of `0.05` Mbps for assertions. This change enhances the flexibility of comparisons for calculated throughput values, allowing for minor variations in results.

t-davidbrown-msft commented on sdk/storage/Azure.Storage.DataMovement/src/ConcurrencyTuner/ThroughputMonitor.cs:
```diff
@@ -0,0 +1,153 @@
+        /// <summary>
+        /// Gets the current throughput in bytes per second.
+        /// </summary>
+        internal virtual decimal Throughput { get; set; }
```
@jalauzon-msft oops. I thought we said double, but I can change them to float.

t-davidbrown-msft commented on sdk/storage/Azure.Storage.DataMovement/src/TransferManager.cs:
```diff
@@ -47,6 +46,46 @@ public class TransferManager : IAsyncDisposable
+                    {
+                        ProgressHandlerOptions = new()
+                        {
+                            TrackBytesTransferred = true
```
@jalauzon-msft Okay. I removed this.

Commit:
Refactor ThroughputMonitor to use float types

Updated the ThroughputMonitor class to change properties and calculations from double to float for improved performance and reduced memory usage.

Changed the access modifier of QueueBytesTransferredAsync from public to internal.

Added a SetStartTime method for direct initialization of _startTime.

Updated tests in ThroughputMonitorTests to reflect these changes, including using float comparisons, renaming tests for clarity, and adding assertions for throughput calculations.
